assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js

```diff
@@ -54,64 +54,81 @@ const usePaymentMethodRegistration = (
 	const canPayArgument = useRef( {
 		cartTotals,
 		cartNeedsShipping,
 		shippingAddress,
+		selectedShippingMethods,
 	} );
 
 	useEffect( () => {
 		canPayArgument.current = {
 			cartTotals,
 			cartNeedsShipping,
 			shippingAddress,
+			selectedShippingMethods,
 		};
-	}, [ cartTotals, cartNeedsShipping, shippingAddress ] );
+	}, [
+		cartTotals,
+		cartNeedsShipping,
+		shippingAddress,
+		selectedShippingMethods,
+	] );
 
-	const resolveCanMakePayments = useCallback( async () => {
-		let initializedPaymentMethods = {},
-			canPay;
-		const setInitializedPaymentMethods = ( paymentMethod ) => {
-			initializedPaymentMethods = {
-				...initializedPaymentMethods,
+	const refreshCanMakePayments = useCallback( async () => {
+		let availablePaymentMethods = {};
+		const addAvailablePaymentMethod = ( paymentMethod ) => {
+			availablePaymentMethods = {
+				...availablePaymentMethods,
 				[ paymentMethod.name ]: paymentMethod,
 			};
 		};
 		for ( const paymentMethodName in registeredPaymentMethods ) {
-			const current = registeredPaymentMethods[ paymentMethodName ];
+			const paymentMethod = registeredPaymentMethods[ paymentMethodName ];
 
+			// In editor, shortcut so all payment methods show as available.
 			if ( isEditor ) {
-				setInitializedPaymentMethods( current );
+				addAvailablePaymentMethod( paymentMethod );
 				continue;
 			}
 
+			// In front end, ask payment method if it should be available.
 			try {
-				canPay = await Promise.resolve(
-					current.canMakePayment( canPayArgument.current )
+				const canPay = await Promise.resolve(
+					paymentMethod.canMakePayment( canPayArgument.current )
 				);
 				if ( canPay ) {
 					if ( canPay.error ) {
 						throw new Error( canPay.error.message );
 					}
-					setInitializedPaymentMethods( current );
+					addAvailablePaymentMethod( paymentMethod );
 				}
 			} catch ( e ) {
+				// If user is admin, show payment `canMakePayment` errors as a notice.
 				handleRegistrationError( e );
 			}
 		}
-		// all payment methods have been initialized so dispatch and set
-		dispatcher( initializedPaymentMethods );
+
+		// Re-dispatch available payment methods to store.
+		dispatcher( availablePaymentMethods );
+
+		// Note: some payment methods use the `canMakePayment` callback to initialize / setup.
+		// Example: Stripe CC, Stripe Payment Request.
+		// That's why we track "is initialised" state here.
 		setIsInitialized( true );
 	}, [ dispatcher, isEditor, registeredPaymentMethods ] );
 
-	// if not initialized invoke the callback to kick off resolving the payments.
+	// Determine which payment methods are available initially and whenever
+	// shipping methods change.
+	// Some payment methods (e.g. COD) can be disabled for specific shipping methods.
 	useEffect( () => {
-		if ( ! isInitialized ) {
-			resolveCanMakePayments();
-		}
-	}, [ resolveCanMakePayments, isInitialized ] );
+		refreshCanMakePayments();
+	}, [ refreshCanMakePayments, selectedShippingMethods ] );
```

Comment:
This `useEffect` is running some times when it's not needed. That's because `selectedShippingMethods` is an object, so when added as an effect dependency, the effect is running again every time because the object reference is always different.

That could be fixed with [`useShallowEqual`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/assets/js/base/hooks/use-shallow-equal.js). Something like:

```JSX
const newSelectedShippingMethods = useShallowEqual(
	selectedShippingMethods
);
```

Then, you could replace `selectedShippingMethods` instances with `newSelectedShippingMethods` and it shouldn't trigger the effect so often.

Reply:
Yes this is a problem. This also changes the logic because `isInitialized` is removed as a dependency.

Here's what I think we should do instead that could also help address some of the point Rua made:

> I'm thinking we could decouple the initialisation of the payment methods from canMakePayment. Not sure if we need to be involved in the init explicitly - maybe payment methods can manage this themselves

There _is_ clear need for payment methods to differentiate between when `canMakePayment` is called as a part of the payment method setup/initialization in checkout and when it's called for other data changes that might cause a change in the availability.  So what I think we should do is:

Introduce a new property on the `canPayArgument.current` of 'isInitialized', and it would be the value of the initialized state.

Keep the _existing_ initialization logic the same (including the `useEffect` here) so that we are preserving the initialization process. So `use-payment-method-registration` hook is explicitly for initialization and only sets the `canPayArgument.current.isInitialized` to true when initialization completes. Payment methods can then use that property to derive when in checkout is in payment method initialization state.

Introduce a _new_ hook that is explicitly for calling `canMakePayment` when various other data (shipping etc) is changed and you feed the hook the `isInitialized` value. So something like this for a signature (name of hook subject to change :) ):

```js
useCanMakePaymentUpdates( isInitialized, paymentMethods, expressPaymentMethods  )
```

Then this hook is used for calling `canMakePayment` on payment methods with the updated data (but will always call with the argument property `isInitialized` true (and only _starts_ calling on data changes when that is true).

Then the hook can be implemented in the same place as `usePaymentMethods` and `useExpressPaymentMethods` are called (`payment-method-data-context.js`) where it receives the `isInitializeValue` returned from those, along with the initialized payment methods from the `paymentData` state in the context. 

A few things will be accomplished by doing this:

- separate initialization logic, from data change/update logic (easier to test and maintain).
- makes the purpose of each hook explicitly clear.
- Payment methods can detect in their `canMakePayment` callback whether they are being initialized by checkout or just being called because of a data update. So it makes it possible for them to decide what needs to happen for the payment method in those different states.